devito/types/separable.py

```diff
@@ -0,0 +1,195 @@
+
+from devito.finite_differences import Differentiable
+from devito.finite_differences.differentiable import Mul, Add
+from devito.tools import as_tuple
+from devito.types.basic import AbstraceCachedAnonymousFunction
+from devito.types.dense import Function, TimeFunction
+
+__all__ = ['SeparableFunction', 'SeparableTimeFunction']
+
+
+class SeparableFunction(Differentiable, AbstraceCachedAnonymousFunction):
+    """
+    A Function separable in its Dimensions, ie f(x,y,z) = f(x)*f(y)*f(z)
+
+    Take the same parameters as a Function plus
+
+    Parameters
+    ----------
+    sep : str
+        Type of separation as a product or sum of Function, 'sum' or 'prod'.
+    separated : Dimension or tuple of Dimensions
+        Dimension in which the Function is separable.
+
+    Examples
+    --------
+    Creation
+
+    >>> from devito import Grid, TimeFunction
+    >>> grid = Grid(shape=(4, 4, 5))
+    >>> x, y, z = grid.dimensions
+    >>> SeparableFunction(name='a', grid=grid)
+    a_x(x)*a_y(y)*a_z(z)
+    >>> SeparableFunction(name='b', grid=grid, op='sum')
+    b_x(x) + b_y(y) + b_z(z)
+    >>> SeparableFunction(name='c', grid=grid, op='sum', separated=x)
+    c_x(x) + c_yz(y, z)
+    """
+    _spearation_op = {'sum': Add, 'prod': Mul}
+
+    def __init__(self, *args, **kwargs):
+        if not self._cached():
+            # Is it sum or product separable
+            self._sep_op = self._spearation_op[kwargs.get('op', 'prod')]
+
+            # Conventional Function inputs
+            self._grid = kwargs.get('grid')
+            self._space_order = kwargs.get('space_order', 1)
+
+            # Initialize subfunctions
+            # Separable dimensions
+            sep_d, nonsep_d = self.__setup_dimensions__(**kwargs)
```

Comment:
argh...just noticed two very important things:

1) 'setup' goes after ! it should have been ``__dimensions_setup__`` ...
2) it's actually ``__indices_setup__`` -- this is how it's called in all other classes of the hierarchy